src/main/java/uk/ac/ebi/ega/egafuse/service/FileChunkDownloadService.java

```diff
@@ -0,0 +1,96 @@
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.google.common.io.CountingInputStream;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import uk.ac.ebi.ega.egafuse.exception.ClientProtocolException;
+import uk.ac.ebi.ega.egafuse.model.CacheKey;
+
+public class FileChunkDownloadService implements IFileChunkDownloadService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileChunkDownloadService.class);
+    private OkHttpClient okHttpClient;
+    private String apiURL;
+    private Token token;
+
+    public FileChunkDownloadService(OkHttpClient okHttpClient, String apiURL, Token token) {
+        this.okHttpClient = okHttpClient;
+        this.apiURL = apiURL;
+        this.token = token;
+    }
+
+    @Retryable(value = {IOException.class, ClientProtocolException.class}, maxAttemptsExpression = "${connection.maxAttempts}", 
+            backoff = @Backoff(delayExpression = "${connection.backoff}"))
+    public byte[] downloadChunk(CacheKey cacheKey) throws IOException, ClientProtocolException {
+        long startCoordinate = cacheKey.getStartCoordinate();
+        long bytesToRead = cacheKey.getChunkBytesToRead();
+       
+        UriComponentsBuilder builder = UriComponentsBuilder.fromPath(apiURL.concat("/files/")).path(cacheKey.getFileId())
+                .queryParam("destinationFormat", "plain").queryParam("startCoordinate", startCoordinate)
+                .queryParam("endCoordinate", (startCoordinate + bytesToRead));
+
+        LOGGER.info("url = " + builder.toUriString());
+
+        Request fileRequest;
+        try {
+            fileRequest = new Request.Builder().url(builder.toUriString()).addHeader("Authorization",
+                    "Bearer " + token.getBearerToken())
+                    .build();
+            try (Response response = okHttpClient.newCall(fileRequest).execute()) {
+                return buildResponseDownloadFiles(response, bytesToRead);
+            } catch (IOException e) {
+                throw new IOException("Unable to execute request. Can be retried.", e);
+            } catch (ClientProtocolException e) {
+                throw new ClientProtocolException(e.toString());
+            }
+        } catch (IOException e) {
+            LOGGER.error("Error in downloading file - {}", e.getMessage(), e);
+            throw new IOException("Unable to execute request. Can be retried.", e);
+        }
+    }
+
+    private byte[] buildResponseDownloadFiles(final Response response, long bytesToRead)
+            throws IOException, ClientProtocolException {
+        final int status = response.code();
+        switch (status) {
+        case 200:
+        case 206:
+            byte[] buffer = new byte[(int) bytesToRead];
+            try (CountingInputStream cIn = new CountingInputStream(response.body().byteStream());

```

Comment:
Why do we need this `CountingInputStream`?

Reply:
Yeah not needed CountingInputStream. Removed now